java-checks/src/main/java/org/sonar/java/checks/spring/NonSingletonAutowiredInSingletonCheck.java

```diff
@@ -0,0 +1,171 @@
+  private static final String AUTOWIRED_ANNOTATION = "org.springframework.beans.factory.annotation.Autowired";
+  private static final Set<String> SINGLETON_LITERALS = Set.of("singleton", "\"singleton\"");
+
+  @Override
+  public List<Tree.Kind> nodesToVisit() {
+    return List.of(Tree.Kind.CLASS);
+  }
+
+  /**
+   * This rule reports an issue when a Singleton Bean auto-wires a non-Singleton Bean via constructor, field or method parameter.
+   */
+  @Override
+  public void visitNode(Tree tree) {
+    ClassTree classTree = (ClassTree) tree;
+
+    if(isSingletonBean(classTree)) {
+      var autowiredBeansStream = Stream.of(
+          autowiredClassFields(classTree),
+          parametersFromAutowiredMethods(classTree),
+          parametersFromAutowiredConstructors(classTree),
+          autowiredParametersFromMethodsAndConstructors(classTree)
+        )
+        .flatMap(stream -> stream)
+        // Remove duplicates in case of both method/constructor and parameters are annotated with @Autowired
+        .distinct();
+
+      analyzeInjectedBeans(autowiredBeansStream);
+    }
+  }
+
+  private static boolean isSingletonBean(ClassTree classTree) {
+    return !hasNotSingletonScopeAnnotation(classTree.modifiers().annotations());
+  }
+
+  private static boolean hasNotSingletonScopeAnnotation(List<AnnotationTree> annotations) {
+    // Only classes annotated with @Scope, having a value different from "singleton", are considered as non-Singleton
+    return annotations.stream().anyMatch(NonSingletonAutowiredInSingletonCheck::isNotSingletonScopeAnnotation);
+  }
+
+  private static boolean isNotSingletonScopeAnnotation(AnnotationTree annotationTree) {
+    return annotationTree.symbolType().is(SCOPED_ANNOTATION)
+      && (isNotSingletonLiteralValue(annotationTree.arguments()) || isNotSingletonAssignmentValue(annotationTree.arguments()));
+  }
+
+  private static boolean isNotSingletonLiteralValue(Arguments arguments){
+    return arguments.size() == 1
+      && arguments.get(0).is(Tree.Kind.STRING_LITERAL)
+      && isNotSingletonLiteral(((LiteralTree) arguments.get(0)).value());
+  }
+
+  private static boolean isNotSingletonAssignmentValue(Arguments arguments) {
+    return arguments
+      .stream()
+      .filter(argument -> argument.is(Tree.Kind.ASSIGNMENT))
+      .map(AssignmentExpressionTree.class::cast)
+      .anyMatch(NonSingletonAutowiredInSingletonCheck::isNotAssignmentToSingletonValue);
+  }
+
+  private static boolean isNotAssignmentToSingletonValue(AssignmentExpressionTree assignmentExpressionTree) {
+    var expression = assignmentExpressionTree.expression();
+
+    if (expression.is(Tree.Kind.STRING_LITERAL)) {
+      return isNotSingletonLiteral(((LiteralTree) expression).value());
+    }
+
+    var variable = (IdentifierTree) assignmentExpressionTree.variable();
+
+    return ("value".equals(variable.name()) || "scopeName".equals(variable.name()))
```

Comment:
I can't figure out how to improve the coverage here :(

Reply:
Consider static imports, which mean it's not a String literal or a member select:
```
import static org.springframework.context.annotation.ScopedProxyMode.TARGET_CLASS;
```